### PR TITLE
Instrument router and remove global spirals (S2S/component)

### PR DESCRIPTION
**Pull request by gustawlippa**
This PR removes global spirals and changes `routingErrors` metric into `router_no_route_found` instrumentation event.

The `router_no_route_found` event is a bit useless, as it will happen if none of the routing modules have routed a stanza (returned `{done, Acc}`), but S2S, as the last router, _always_ tries to route a message and returns `done`. If it can't find the recipient, it sends the error stanzas on its own. So for this event to occur, S2S has to be disabled - and this is how I tested it.

I wasn't sure how to handle the metrics to be similar to C2S ones. There are two issues that we can still decide to change in this PR:
- naming: I went with the `_in/_out` convention from #4310, but transport metrics for C2S,(which are more similar to the ones from this PR) use `_received/_sent` convention from  #4304. I think it's best to either change them to use `_in/_out` as well, or I can change the names to `_received/_sent` in this PR.
- number of metrics: I decided for this PR, to change the metrics one for one, but there could be more metrics, to mirror the ones for C2S. I created a PR with a proposed implementation of this: https://github.com/esl/MongooseIM/pull/4322. However, I'm not certain if we need so many additional metrics. Additional idea is that all of them could be combined, and just have labels `c2s/s2s/component`.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/227075) / elasticsearch_and_cassandra_mnesia / 582afa83980b60669ffb936271a93c82403efb51
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/227075/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/227075/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40b5937be5397a.2024-07-08_11.53.23/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/227305) / elasticsearch_and_cassandra_mnesia / f8ae4b690a3d4b7a9d6422d6a68d4aceed17450e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/227305/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/227305/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4099871bb2a2d5.2024-07-10_09.43.34/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/227324) / elasticsearch_and_cassandra_mnesia / 980ee28952eeeb29827458b67450c9336ca9381c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/227324/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/227324/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4061a73fa59116.2024-07-10_09.43.36/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/227350) / elasticsearch_and_cassandra_mnesia / 6ca9c020cb1f3a7a889dd3f6e3a935ef136057e0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/227350/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/227350/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4008af8b8e3c76.2024-07-10_12.28.43/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/227647) / elasticsearch_and_cassandra_mnesia / cde4b6a4e4a632985abb9ff2766b243b1d952f5a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/227647/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/227647/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%401a8574c0f134.2024-07-11_08.51.50/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Review by chrzaszcz — COMMENTED** [review]
Looks good in general. I added a few minor questions and notes.

File: doc/operation-and-maintenance/MongooseIM-metrics.md [hunk 1]
```diff
@@ -182,23 +182,23 @@ Metrics specific to an extension, e.g. Message Archive Management, are described
...
-| `[global, data, xmpp, sent, s2s]` | spiral | A size (in bytes) of a data sent via TCP and TLS (before encryption) Server-to-Server connections. |
-| `[global, data, xmpp, received, component]` | spiral | A size (in bytes) of a data received from XMPP component. |
-| `[global, data, xmpp, sent, component]` | spiral | A size (in bytes) of a data sent to XMPP component. |
-| `[global, data, dist]` | proplist | Network stats for an Erlang distributed communication. A proplist with values: `recv_oct`, `recv_cnt`, `recv_max`, `send_oct`, `send_max`, `send_cnt`, `send_pend`, `connections`. |
+| Metric name                                        | Type      | Description                                                                                                                                                                        |
+|----------------------------------------------------|-----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply doc/operation-and-maintenance/MongooseIM-metrics.md:186]: I will try to revert it - my editor keeps "improving" the whole table the moment I touch a row. It's annoying for me as well.

File: src/mongoose_transport.erl [hunk 2]
```diff
@@ -389,37 +387,39 @@ deactivate_socket(#state{socket = Socket, sockmod = mongoose_tls}) ->
...
         end,
     {NewShaperState, Pause} = mongoose_shaper:update(ShaperState, Size),
-    update_transport_metrics(Size, received, State#state.connection_type),
+    update_transport_metrics(Data, in, ConnectionType),
     [gen_fsm_compat:send_event(DestPid, Event) || Event <- Events],
     maybe_pause(Pause, State),
     State#state{parser = NewParser, shaper_state = NewShaperState}.
 
-wrap_xml_elements_and_update_metrics(#xmlel{} = E) ->
-    mongoose_instrument:execute(c2s_tcp_data_received, #{}, #{byte_size => exml:xml_size(E)}),
```
[the remark this answers is not in the document]
Comment (gustawlippa) [reply src/mongoose_transport.erl:409]: Yes, I think so. C2S doesn't use `mongoose_transport` at all - it has it's own modules for sockets, the main one being `mongoose_c2s_ranch`.
There may have been some dynamic calls that I've missed, but I don't think so. This metric was not tested (removing it here didn't break any tests), so I think it was an oversight during the C2S data transport metrics rework.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/228166) / elasticsearch_and_cassandra_mnesia / 88032ac5df707812bb7ea85cfa7e25a3974613d5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/228166/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/228166/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4063a12a4878e2.2024-07-12_14.01.37/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/228238) / elasticsearch_and_cassandra_mnesia / 9fbd05d400ddec5d075e941c3ac29a13aa90cab9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/228238/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/228238/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4031023de1004e.2024-07-12_15.49.52/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/228904) / elasticsearch_and_cassandra_mnesia / 87ee12e36683a9bf3b610c1d832d8ad98d783e5f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/228904/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/228904/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%403af14885736f.2024-07-15_14.28.45/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/228945) / elasticsearch_and_cassandra_mnesia / d85d90aeba9b7702c585b282c4e5781ebce40781
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/228945/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/228945/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40e344012592c8.2024-07-15_14.43.32/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/228974) / elasticsearch_and_cassandra_mnesia / 118bf12ba77dafcd6170b04fd2a6cc427a69ab52
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/228974/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/228974/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40e889e8bf7401.2024-07-15_14.44.29/index.html)
**OK: 456** / Failed: 0 / **User-skipped: 45** / **Auto-skipped: 1** 
<details><summary>mam_SUITE:cassandra_muc_prefs_cases:muc_messages_filtered_when_prefs_default_policy_is_never</summary>


```erlang
{skip,
  {failed,
    {mam_SUITE,init_per_testcase,
      {{badrpc,
         {'EXIT',
           {#{module => mod_mam_cassandra_prefs,
            text => <<"Module missing from mongoose_config">>,
            host_type => <<"localhost">>,
            what => module_not_loaded},
            [{gen_mod,assert_loaded,2,
               [{file,"/home/circleci/project/src/gen_mod.erl"},
              {line,346}]},
             {gen_mod,start_module,3,
               [{file,"/home/circleci/project/src/gen_mod.erl"},
              {line,96}]},
             {mongoose_modules,start_module,4,
               [{file,
                "/home/circleci/project/src/mongoose_modules.erl"},
              {line,90}]},
             {mongoose_modules,ensure_started,3,
               [{file,
                "/home/circleci/project/src/mongoose_modules.erl"},
              {line,80}]},
             {mongoose_modules,'-replace_modules/3-lc$^1/1-1-',2,
               [{file,
                "/home/circleci/project/src/mongoose_modules.erl"},
              {line,50}]},
             {mongoose_modules,replace_modules,3,
               [{file,
                "/home/circleci/project/src/mongoose_modules.erl"},
              {line,51}]}]}}},
       [{distributed_helper,rpc,
          [#{node => mongooseim@localhost},
           mongoose_modules,replace_modules,
           [<<"localhost">>,[],
          #{mod_mam =>
              #{cache =>
                #{module => internal,strategy => fifo,
             ...
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/228974/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40e889e8bf7401.2024-07-15_14.44.29/big_tests.tests.mam_SUITE.logs/run.2024-07-15_14.44.31/mam_suite.muc_messages_filtered_when_prefs_default_policy_is_never.12546.html)
</details>

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229046) / elasticsearch_and_cassandra_mnesia / ece35360ac1834f172ad4f8fdd17afba851622d2
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229046/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229046/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%406d5ad1589533.2024-07-15_16.47.13/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229087) / elasticsearch_and_cassandra_mnesia / bff7daa124b3934011842d9f98e89ddcfdf5c11a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229087/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229087/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40e3415901a6b7.2024-07-15_17.12.15/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229110) / elasticsearch_and_cassandra_mnesia / cf1c165788f35f73cf8ba2b99dad8fe00d29035f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229110/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229110/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40341b66a90765.2024-07-15_17.47.35/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229366) / elasticsearch_and_cassandra_mnesia / cdfa58ea7a5cb253cac09c6286b538348cf54cb8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229366/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229366/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40cb73f603ff44.2024-07-16_10.47.59/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229406) / elasticsearch_and_cassandra_mnesia / f193cb01266391085fdb72d074e9df99d2fee8e6
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229406/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229406/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40818d2a9ec398.2024-07-16_12.00.12/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229435) / elasticsearch_and_cassandra_mnesia / 30e684c476b4e795dd8d2c06c258bffef3b0a500
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229435/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229435/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%404da4e71f1f5b.2024-07-16_12.00.52/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229458) / elasticsearch_and_cassandra_mnesia / fe995bce9e620125ed0f334a444d949c991a8f20
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229458/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229458/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40146f0433fbb5.2024-07-16_13.42.34/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229491) / elasticsearch_and_cassandra_mnesia / 6d28edfc34a333205e2a0ad9c642c369ad915584
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229491/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229491/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4008b3b15c8276.2024-07-16_14.18.30/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/229532) / elasticsearch_and_cassandra_mnesia / 578bd1b93a65bf3d401197951eba7965f6f6ff4b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4319/229532/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4319/229532/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4009279a59eb0e.2024-07-16_15.27.14/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 45** / Auto-skipped: 0

**Review by JanuszJakubiec — APPROVED** [review]
Looks good

**Pull request merged**